# [E-01] (2) load_wine : 와인을 분류해 봅시다

In [1]:
# 라이브러리 버전 확인
import sklearn

print(sklearn.__version__)

1.0


In [2]:
# 필요한 모듈 import하기
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

In [3]:
# 데이터 준비
wine = load_wine()
wine_data = wine.data
wine_label = wine.target

In [4]:
wine.keys() # wine에 어떤 정보가 담겼는지 확인

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names'])

In [5]:
print(wine_data.shape) # 데이터의 크기 확인

(178, 13)


In [6]:
wine_data[0] # 샘플로 하나의 데이터만 확인

array([1.423e+01, 1.710e+00, 2.430e+00, 1.560e+01, 1.270e+02, 2.800e+00,
       3.060e+00, 2.800e-01, 2.290e+00, 5.640e+00, 1.040e+00, 3.920e+00,
       1.065e+03])

In [7]:
# 특성들의 이름 확인
wine.feature_names

['alcohol',
 'malic_acid',
 'ash',
 'alcalinity_of_ash',
 'magnesium',
 'total_phenols',
 'flavanoids',
 'nonflavanoid_phenols',
 'proanthocyanins',
 'color_intensity',
 'hue',
 'od280/od315_of_diluted_wines',
 'proline']

In [8]:
# 특성 갯수 확인 
len(wine.feature_names)

13

In [9]:
# target 데이터 확인
print(wine_label.shape)
wine_label[:20]

(178,)


array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [10]:
# 분류하고자 하는 대상 확인
print(wine.target_names)

['class_0' 'class_1' 'class_2']


In [11]:
# wine 데이터에 대한 설명
print(wine.DESCR)

.. _wine_dataset:

Wine recognition dataset
------------------------

**Data Set Characteristics:**

    :Number of Instances: 178 (50 in each of three classes)
    :Number of Attributes: 13 numeric, predictive attributes and the class
    :Attribute Information:
 		- Alcohol
 		- Malic acid
 		- Ash
		- Alcalinity of ash  
 		- Magnesium
		- Total phenols
 		- Flavanoids
 		- Nonflavanoid phenols
 		- Proanthocyanins
		- Color intensity
 		- Hue
 		- OD280/OD315 of diluted wines
 		- Proline

    - class:
            - class_0
            - class_1
            - class_2
		
    :Summary Statistics:
    
    ============================= ==== ===== ======= =====
                                   Min   Max   Mean     SD
    ============================= ==== ===== ======= =====
    Alcohol:                      11.0  14.8    13.0   0.8
    Malic Acid:                   0.74  5.80    2.34  1.12
    Ash:                          1.36  3.23    2.36  0.27
    Alcalinity of Ash:            1

In [12]:
# 특성 행렬을 pandas의 DataFrame으로 나타내기
import pandas as pd

pd.DataFrame(wine.data, columns=wine.feature_names)

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0      14.23        1.71  2.43               15.6      127.0           2.80   
1      13.20        1.78  2.14               11.2      100.0           2.65   
2      13.16        2.36  2.67               18.6      101.0           2.80   
3      14.37        1.95  2.50               16.8      113.0           3.85   
4      13.24        2.59  2.87               21.0      118.0           2.80   
..       ...         ...   ...                ...        ...            ...   
173    13.71        5.65  2.45               20.5       95.0           1.68   
174    13.40        3.91  2.48               23.0      102.0           1.80   
175    13.27        4.28  2.26               20.0      120.0           1.59   
176    13.17        2.59  2.37               20.0      120.0           1.65   
177    14.13        4.10  2.74               24.5       96.0           2.05   

     flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0          3.06                  0.28             2.29             5.64  1.04   
1          2.76                  0.26             1.28             4.38  1.05   
2          3.24                  0.30             2.81             5.68  1.03   
3          3.49                  0.24             2.18             7.80  0.86   
4          2.69                  0.39             1.82             4.32  1.04   
..          ...                   ...              ...              ...   ...   
173        0.61                  0.52             1.06             7.70  0.64   
174        0.75                  0.43             1.41             7.30  0.70   
175        0.69                  0.43             1.35            10.20  0.59   
176        0.68                  0.53             1.46             9.30  0.60   
177        0.76                  0.56             1.35             9.20  0.61   

     od280/od315_of_diluted_wines  proline  
0                            3.92   1065.0  
1                            3.40   1050.0  
2                            3.17   1185.0  
3                            3.45   1480.0  
4                            2.93    735.0  
..                            ...      ...  
173                          1.74    740.0  
174                          1.56    750.0  
175                          1.56    835.0  
176                          1.62    840.0  
177                          1.60    560.0  

[178 rows x 13 columns]

In [13]:
# 데이터 분리
x_train, x_test, y_train, y_test = train_test_split(wine_data,
                                                    wine_label,
                                                    test_size=0.2,
                                                    random_state=30)

In [14]:
# 모델 학습 및 예측
##Decision Tree 사용
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

decision_tree = DecisionTreeClassifier()
decision_tree.fit(x_train,y_train)
y_pred = decision_tree.predict(x_test)

print(classification_report(y_test, y_pred))
print("정답률=", accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.83      0.91      0.87        11
           1       0.87      0.81      0.84        16
           2       0.89      0.89      0.89         9

    accuracy                           0.86        36
   macro avg       0.86      0.87      0.87        36
weighted avg       0.86      0.86      0.86        36

정답률= 0.8611111111111112


In [15]:
## Ransom Forest 사용
from sklearn.ensemble import RandomForestClassifier

random_forest = RandomForestClassifier()
random_forest.fit(x_train, y_train)
y_pred = random_forest.predict(x_test)

print(classification_report(y_test, y_pred))
print("정답률=", accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.92      1.00      0.96        11
           1       1.00      0.94      0.97        16
           2       1.00      1.00      1.00         9

    accuracy                           0.97        36
   macro avg       0.97      0.98      0.97        36
weighted avg       0.97      0.97      0.97        36

정답률= 0.9722222222222222


In [16]:
## Support Vector Machine (SVM) 사용
from sklearn import svm
svm_model = svm.SVC()

svm_model.fit(x_train, y_train)
y_pred = svm_model.predict(x_test)

print(classification_report(y_test, y_pred))
print("정답률=", accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.89      0.73      0.80        11
           1       0.79      0.69      0.73        16
           2       0.46      0.67      0.55         9

    accuracy                           0.69        36
   macro avg       0.71      0.69      0.69        36
weighted avg       0.74      0.69      0.71        36

정답률= 0.6944444444444444


In [17]:
## Stochastic Gradient Descent Classifier (SGDClassifier) 사용
from sklearn.linear_model import SGDClassifier
sgd_model = SGDClassifier()

sgd_model.fit(x_train, y_train)
y_pred = sgd_model.predict(x_test)

print(classification_report(y_test, y_pred))
print("정답률=", accuracy_score(y_test, y_pred))
# 가끔 0,00이 뜸. 램덤한 값에서 예측하기 때문에 가끔 그럼.

              precision    recall  f1-score   support

           0       1.00      0.55      0.71        11
           1       0.00      0.00      0.00        16
           2       0.30      1.00      0.46         9

    accuracy                           0.42        36
   macro avg       0.43      0.52      0.39        36
weighted avg       0.38      0.42      0.33        36

정답률= 0.4166666666666667


/opt/conda/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1308: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/conda/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1308: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/conda/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1308: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [18]:
## Logistic Regression 사용
from sklearn.linear_model import LogisticRegression
logistic_model = LogisticRegression()

logistic_model.fit(x_train, y_train)
y_pred = logistic_model.predict(x_test)

print(classification_report(y_test, y_pred))
print("정답률=", accuracy_score(y_test, y_pred))
# digits.data를 할때와 마찬가지. 밑에 코드에서 max 값을 더 크게 줌(5000)

              precision    recall  f1-score   support

           0       1.00      0.91      0.95        11
           1       0.94      1.00      0.97        16
           2       1.00      1.00      1.00         9

    accuracy                           0.97        36
   macro avg       0.98      0.97      0.97        36
weighted avg       0.97      0.97      0.97        36

정답률= 0.9722222222222222


/opt/conda/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [19]:
from sklearn.linear_model import LogisticRegression
logistic_model = LogisticRegression(max_iter=5000)

logistic_model.fit(x_train, y_train)
y_pred = logistic_model.predict(x_test)

print(classification_report(y_test, y_pred))
print("정답률=", accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.92      1.00      0.96        11
           1       1.00      0.94      0.97        16
           2       1.00      1.00      1.00         9

    accuracy                           0.97        36
   macro avg       0.97      0.98      0.97        36
weighted avg       0.97      0.97      0.97        36

정답률= 0.9722222222222222


# 모델 평가

- Decision Tree
    정답률 = 0.86
- Random Forest
    정답률 = 0.97 (0.972222..)
- Support Vector Machine (SVM)
    정답률 = 0.69
- Stochastic Gradient Descent Classifier (SGDClassifier)
    정답률 = 0.42
- Logistic Regression
    정답률 = 0.97 (0.97222...)
    
Random Forest와 Logistic Regression의 정답률이 똑같음. 그러므로 둘다 wine_data를 잘 구분하는 듯.

# 회고

- 알게 된 점 :
    
    Stochastic Gradient Descent Classifier (SGDClassifier) 사용할 때 가끔 0.00이 뜸. 확인해보니 램덤한 값에서 평균을 내다보니 가끔 0.00이 뜰 수 있음. 다시 실행해보면 괜찮아짐.
    

- digits를 하고 해보니 좀 더 이해가 잘되는 느낌.